File: src/NodeParser/ArrayNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import { AnyType } from "../Type/AnyType.js";
import { ArrayType } from "../Type/ArrayType.js";
import type { BaseType } from "../Type/BaseType.js";

export class ArrayNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.ArrayTypeNode): boolean {
        return node.kind === ts.SyntaxKind.ArrayType;
    }

    public createType(node: ts.ArrayTypeNode, context: Context): BaseType {
        const type = this.childNodeParser.createType(node.elementType, context);
        // Generics without `extends` or `defaults` cannot be resolved, so we fallback to `any`
        return new ArrayType(type ?? new AnyType());
    }
}

File: internal/parser/array_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ArrayNodeParser handles array type nodes like `string[]`
// (src/NodeParser/ArrayNodeParser.ts).
type ArrayNodeParser struct {
	childNodeParser NodeParser
}

func NewArrayNodeParser(childNodeParser NodeParser) *ArrayNodeParser {
	return &ArrayNodeParser{childNodeParser: childNodeParser}
}

func (p *ArrayNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindArrayType
}

func (p *ArrayNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	typ := p.childNodeParser.CreateType(node.AsArrayTypeNode().ElementType, ctx, nil)
	// Generics without `extends` or `defaults` cannot be resolved, so we fall back to `any`.
	if typ == nil {
		typ = &types.AnyType{}
	}
	return &types.ArrayType{Item: typ}
}
